cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/OAuthHandler.java

```
@@ -264,9 +265,29 @@ public void getOAuthCredential(HttpServiceRequest request, HttpServiceResponder
       } catch (JsonSyntaxException e) {
         throw new OAuthServiceException(HttpURLConnection.HTTP_INTERNAL_ERROR, "Error parsing JSON response", e);
       }
-      if (refreshTokenResponse.getAccessToken() == null || refreshTokenResponse.getAccessToken().isEmpty()) {
+
+      boolean hasRefreshToken = refreshTokenResponse.getRefreshToken() != null
+          && !refreshTokenResponse.getRefreshToken().isEmpty();
+      boolean hasAccessToken = refreshTokenResponse.getAccessToken() != null
+          && !refreshTokenResponse.getAccessToken().isEmpty();
+
+      if (!hasAccessToken) {
         throw new OAuthServiceException(
-            HttpURLConnection.HTTP_INTERNAL_ERROR, "Refresh token response body does not have refresh token");
+            HttpURLConnection.HTTP_INTERNAL_ERROR, "Access token response body does not have access token");
+      }
+
+      // API has given us a new refresh token
+      if (hasRefreshToken && !refreshToken.getRefreshToken().equals(refreshTokenResponse.getRefreshToken())) {
+        OAuthRefreshToken newRefreshToken = OAuthRefreshToken.newBuilder()
+                .withRefreshToken(refreshTokenResponse.getRefreshToken())
+                .withRedirectURI(refreshToken.getRedirectURI())
+                .build();
+
+        try {
+          oauthStore.writeRefreshToken(provider, credentialId, newRefreshToken);
+        } catch (OAuthStoreException e) {
+          LOG.error("An error occurred while writing the new refresh token", e);
```

I am not sure if we should just ignore the error, what if the older refresh_token got corrupted, I would suggest throwing a `5xx`.

---

If the old token got corrupted somehow I would think the overwrite should still succeed, no?
For now I've changed this to throw 500.